### ISSUE TITLE
Bare Metal Cilium l2 Advertisement not sending ARPS

### ISSUE THREAD

**Issue by errolsancaktar** [cilium/cilium#34835]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

equal or higher than v1.16.0 and lower than v1.17.0

### What happened?

Installed cilium on a new cluster in bare metal.  Everything runs and l2 advertisement seems to work assigning an ip that is accessible from the node.  it is not accessible from outside the node however.  the logs dont seem to show any ARP notices either which  is confusing.  

### How can we reproduce the issue?

helm install cilium cilium/cilium --version 1.16.1 -f components/cilium.yaml --namespace kube-system

cilium.yaml:
k8sServiceHost: "10.10.9.11"
k8sServicePort: "6443"
kubeProxyReplacement: true
debug:
  enabled: true
bpf:
  enabled: true
  direct-routing-device: "eth0"
  masquerade: true
ipMasqAgent:
  enabled: true
l2announcements:
  enabled: true
l2podAnnouncement:
  interface: "eth0"
  enabled: true
ingressController:
  enabled: true
  default: true
  loadbalancerMode: shared
  enforceHttps: false
  enableProxyProtocol: false
  hostNetwork:
    enable: true
  service:
    type: NodePort
hostPort:
  enabled: true
externalIPs:
  enabled: true
hubble:
  enabled: true
  relay:
    enabled: true
ipam:
  mode: "cluster-pool"
  ciliumNodeUpdateRate: "15s"
  operator:
    clusterPoolIPv4PodCIDRList: ["10.244.0.0/16"]
    clusterPoolIPv4MaskSize: 24
logSystemLoad: true
requireIPv4PodCIDR: true
egressGateway:
  enabled: false
loadBalancer:
  experimental: false
nodePort:
  enabled: true
operator:
  enabled: true
  rollOutPods: true
  replicas: 1
rollOutCiliumPods: true

### Cilium Version

1.16.1

### Kernel Version

Linux node1 6.8.0-44-generic #44-Ubuntu SMP PREEMPT_DYNAMIC Tue Aug 13 13:35:26 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

1.31.0

### Sysdump

[cilium-sysdump-20240911-173712.zip](https://github.com/user-attachments/files/16968675/cilium-sysdump-20240911-173712.zip)

### Anything else?

root@node1:~/kubernetes/install# kubectl get pods -n kube-system
NAME                               READY   STATUS    RESTARTS       AGE
cilium-7pg4x                       1/1     Running   0              55m
cilium-envoy-n98xm                 1/1     Running   6 (113m ago)   3h13m
cilium-operator-569686fc7d-22sbf   1/1     Running   0              55m
coredns-6f6b679f8f-n7w6x           1/1     Running   2 (113m ago)   3h15m
coredns-6f6b679f8f-qj4hx           1/1     Running   2 (113m ago)   3h15m
etcd-node1                         1/1     Running   3 (113m ago)   3h15m
hubble-relay-5fc44b85db-vsczv      1/1     Running   8 (91m ago)    3h13m
kube-apiserver-node1               1/1     Running   3 (113m ago)   3h15m
kube-controller-manager-node1      1/1     Running   8 (55m ago)    3h15m
kube-scheduler-node1               1/1     Running   3 (113m ago)   3h15m

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by dylandreimerink** [comment]
Did you create any `CiliumL2AnnouncementPolicy`s? What do they look like?  You can also try running through the troubleshooting guide to find common issues https://docs.cilium.io/en/latest/network/l2-announcements/#troubleshooting

**Comment by JamesLavin** [comment]
I just upgraded from Cilium 1.15.7 to 1.16.1, and my L2 announcements seemed to stop working. I previously had an assigned and functional LoadBalancer External IP under both 1.14 and 1.15. It stopped working when I upgraded to 1.16, and I can no longer access my websites from outside my cluster (though I can still access them via NodePorts):
```
bjorn@gustavsson:~/Helm/cilium$ kubectl get svc -A
NAMESPACE      NAME                          TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)                      AGE
cert-manager   cert-manager                  ClusterIP      10.107.220.68    <none>        9402/TCP                     258d
cert-manager   cert-manager-webhook          ClusterIP      10.96.96.120     <none>        443/TCP                      258d
cnpg-system    cnpg-webhook-service          ClusterIP      10.104.204.85    <none>        443/TCP                      195d
default        cilium-gateway-http-gateway   LoadBalancer   10.111.148.3     <pending>     80:32078/TCP                 208d
default        cilium-gateway-tls-gateway    LoadBalancer   10.102.131.115   <pending>     443:32532/TCP                206d
default        [name_censored]               LoadBalancer   10.105.139.55    <pending>     4000:31838/TCP               262d
...
```

Everything still seems configured properly:
```
bjorn@gustavsson:~/Helm/cilium$ kubectl -n kube-system exec ds/cilium -- cilium-dbg config --all | grep EnableL2Announcements
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
EnableL2Announcements             : true
bjorn@gustavsson:~/Helm/cilium$ kubectl -n kube-system exec ds/cilium -- cilium-dbg config --all | grep KubeProxyReplacement
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
KubeProxyReplacement              : true
KubeProxyReplacementHealthzBindAddr: 
bjorn@gustavsson:~/Helm/cilium$ kubectl -n kube-system exec ds/cilium -- cilium-dbg config --all | grep EnableExternalIPs
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
EnableExternalIPs                 : true
bjorn@gustavsson:~/Helm/cilium$ kubectl get CiliumL2AnnouncementPolicy
NAME                     AGE
[censored-name]-policy   207d
```

Probably unrelated but possibly worth noting: I see in the release notes the following, which I don't fully understand:
> L7 network policy with terminatingTLS will not load the key ca.crt even if it is present in the secret. This prevents Envoy from incorrectly requiring client certificates from pods when using TLS termination. To retain old behaviour for bug compatibility, please set --use-full-tls-context=true.

I updated via Helm twice, first without that setting and a second time with it. Neither time could I access my cluster from externally or get a LoadBalancer External IP address.

**Comment by JamesLavin** [comment]
When I tried rolling back via Helm to 1.15, my cilium-operator pods got stuck. I looked at the logs and see it complaining about a missing Gateway API CRD:
```
time="2024-09-14T19:30:42Z" level=info msg="Checking for required GatewayAPI resources" requiredGVK="[gateway.networking.k8s.io/v1, Kind=gatewayclasses gateway.networking.k8s.io/v1, Kind=gateways gateway.networking.k8s.io/v1, Kind=httproutes gateway.networking.k8s.io/v1beta1, Kind=referencegrants gateway.networking.k8s.io/v1alpha2, Kind=grpcroutes gateway.networking.k8s.io/v1alpha2, Kind=tlsroutes]" subsys=gateway-api
time="2024-09-14T19:30:42Z" level=error msg="Invoke failed" ="gateway-api.initGatewayAPIController (pkg/gateway-api/cell.go:70)" error="failed to create gateway controller: failed to setup reconciler: no matches for kind \"GRPCRoute\" in version \"gateway.networking.k8s.io/v1alpha2\"" subsys=hive
time="2024-09-14T19:30:42Z" level=info msg=Stopping subsys=hive
time="2024-09-14T19:30:42Z" level=fatal msg="failed to start: failed to create gateway controller: failed to setup reconciler: no matches for kind \"GRPCRoute\" in version \"gateway.networking.k8s.io/v1alpha2\"" subsys=cilium-operator-generic
```

GRPCRoute (which I'm not even using, though I am using Gateway API... or was until upgrading to Cilium 1.16.1 broke it) got upgraded recently from experimental channel to standard channel, and that has apparently messed me up when trying to roll back from Cilium 1.16.1 back to 1.15. https://gateway-api.sigs.k8s.io/guides/#v11-upgrade-notes says:
> With the graduation of GRPCRoute to GA, it is now included in standard channel. Unfortunately, that can be problematic for anyone that was already using the experimental channel version of GRPCRoute. As a rule, CRDs in standard channel do not expose alpha API version to avoid any version deprecations in that channel. That means that the standard channel version of GRPCRoute excludes v1alpha2. All implementations of GRPCRoute built before the v1.1 release of Gateway API would have exclusively relied on v1alpha2 and will need to be updated to support GRPCRoute v1. Until implementations have been updated to support v1, you can safely upgrade to the experimental channel version of GRPCRoute included in v1.1 that exposes both v1 and v1alpha2.
> 
> Upgrade Sequence: If you're already using v1alpha2 GRPCRoute, we'd recommend the following upgrade sequence:
> 
> Install experimental v1.1 GRPCRoute CRD
> Update all your manifests to use v1 instead of v1alpha2
> Upgrade to an implementation that supports GRPCRoute v1 API Version
> Install standard channel v1.1 GRPCRoute CRD

Deep sigh.

**Comment by JamesLavin** [comment]
Gateway API versioning is damn confusing: https://github.com/kubernetes-sigs/gateway-api/pull/3084/files. (I'm uncertain this is the OP's problem, but mentioning since it seems relevant to my messed up cluster and I _think_ I hit the same problem.)

**Comment by JamesLavin** [comment]
I hope this info is relevant. Apologies if it's not...

I couldn't even add the Gateway API CRD my failing cilium-operators couldn't find:
```
bjorn@gustavsson:~/Helm/Gateway_API$ kubectl apply -f v1.0.0/
customresourcedefinition.apiextensions.k8s.io/gatewayclasses.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/gateways.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/httproutes.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/referencegrants.gateway.networking.k8s.io configured
customresourcedefinition.apiextensions.k8s.io/tlsroutes.gateway.networking.k8s.io configured
The CustomResourceDefinition "grpcroutes.gateway.networking.k8s.io" is invalid: status.storedVersions[1]: Invalid value: "v1": must appear in spec.versions
```

I finally got my Cilium operators running by deleting my API Gateways and then deleting my Gateway API CRDs.

Once cilium settles down, I'll try recreating all that.

**Comment by JamesLavin** [comment]
FWIW, I just got my websites back online. Sharing how I did it in case this is helpful info for anyone else...

Cilium looked fine when I stepped stepped away this afternoon, but no my cluster was externally inaccessible.

When I returned tonight, I first tried to fix Gateway API, my L2 announcements, and IPAM. I immediately noticed there were no IPs available in my CiliumLoadBalancerIPPool:
```
bjorn@gustavsson:~$ kubectl get ciliumloadbalancerippool -A
NAME        DISABLED   CONFLICTING   IPS AVAILABLE   AGE
main-pool   false      False         0               216d
```
I fixed this by deleting `main-pool` and recreating it. I got a syntax error and had to replace `cidrs:` with `blocks:` in my YAML file (perhaps the syntax changed at some point?). I then saw that my only LoadBalancer service now had an external IP assigned. I thought the last step in fixing my setup (after reverting to Cilium 1.15 from Cilium 1.16.1) was just remembering how I got an external IP assigned to my Gateway API gateway, but that actually wasn't my problem.

I discovered that only one of my two cilium-operators had successfully updated. The logs of the other was also complaining about the same grpcroutes problem above. So I re-deleted my Gateway and the Gateway API CRDs and eventually got the second cilium-operator to start, but I noticed that none of the other Cilium pods had restarted since I ran "helm rollback." I found some config settings in https://blog.stonegarden.dev/articles/2023/12/cilium-gateway-api/#summary that I had never set (or heard of) previously that sounded potentially relevant, so I tweaked my Helm YAML and tried upgrading to Cilium 1.16.1 again.

This upgrade seemed to work... I saw ALL the Cilium pods restart this time. (I also have been fighting complete node network loss every time I upgrade Cilium due to a Cilium/CRI-O bug that was finally fixed in 1.16, but I'm painfully familiar with how to manually re-enable each node's networking.) Then I was able to re-apply my gateway YAML, and everything started working. My Gateway now has a different external IP than it had previously, but I pointed HAProxy at the new IP and now everything's good again.

Good luck to anyone else who experiences similar issues upgrading to 1.16. I hope some clue from my experience proves helpful. My main suggestion would be to just delete your Gateway API(s), Gateway API CRDs, and possibly CiliumLoadBalancerIPPools, etc., then re-apply them after upgrading to Cilium 1.16.1. That's what seemed to work for me.
